Add blog about repo moves. (#8)
CHERIoT-Platform/CHERIoT-Platform.github.io: _posts/2024-07-31-moving-to-the-cheriot-org.markdown

---
layout: post
title:  "CHERIoT projects moving into the CHERIoT Platform org"
# Fix this and the file name before publishing!
date:   2024-07-31
categories: rtos sail
author: David Chisnall
---

> I am please to announce that, today, Microsoft has transferred the [CHERIoT Sail](https://github.com/CHERIoT-Platform/cheriot-sail) and [CHERIoT RTOS](https://github.com/CHERIoT-Platform/cheriot-rtos) repositories into the CHERIoT-Platform GitHub organisation.

Reviewer: ```suggestion
I am pleased to announce that, today, Microsoft has transferred the [CHERIoT Sail](https://github.com/CHERIoT-Platform/cheriot-sail) and [CHERIoT RTOS](https://github.com/CHERIoT-Platform/cheriot-rtos) repositories into the CHERIoT-Platform GitHub organisation.
```

CHERIoT began as a research project in Microsoft Research Cambridge, as part of Microsoft's work on the [Digital Security by Design Programme](https://www.dsbd.tech).
The project's goal was to explore several aspects of CHERI system design, specifically:

 - What can you do with software abstractions if you can assume CHERI?
 - How can you scale down the work on temporal safety to tiny embedded devices?
 - How can you remove the need for bolted-on security extensions by creating a holistic hardware-software security model around an assumption of memory safety?

As a research project, it was a resounding success.

> We demonstrated that a foundation of CHERI let you build tiny cores that provided a step change in the baseline security guarantees, along with a simple programmer model, in exchange for a tiny power and area overhead.

Reviewer: ```suggestion
[We demonstrated](https://www.microsoft.com/en-us/research/publication/cheriot-complete-memory-safety-for-embedded-devices/) that a foundation of CHERI let you build tiny cores that provided a step change in the baseline security guarantees, along with a simple programmer model, in exchange for a tiny power and area overhead.
```

> Microsoft; however, is not a microcontroller vendor and, for CHERIoT to be useful, it needed broad ecosystem support.

Reviewer: ```suggestion
Microsoft; however, is not a microcontroller vendor and, for CHERIoT to be useful, it needs broad ecosystem support.
```

> This ecosystem is starting to grow.

Reviewer: That feels a little terse... maybe something like

```suggestion
This ecosystem is forming, with several companies making significant early contributions.
```

> We have merged improvements to the ISA specification and various parts of the software stack from Google.

Reviewer: Front Google in the sentence like in the others?
```suggestion
Google has contributed, and we have merged, improvements to the ISA specification and various parts of the software stack.
```

SCI Semiconductor is working to ship commercial CHERIoT SoCs to customers next year.
lowRISC has built the [Sonata FPGA platform](https://www.sunburst-project.org) for prototyping CHERIoT devices.
Folks at [Configured Things](https://www.configuredthings.com) have written a [fantastic demo application](https://github.com/CHERIoT-Platform/cheriot-demos/tree/main/configuration_broker)

> At this point, CHERIoT is no longer a research project, it is an open source foundation supported by multiple vendors and the repository moves reflect this fact.

Reviewer: ```suggestion
At this point, CHERIoT is no longer a research project, it is an open source foundation supported by multiple vendors and these repositories' moves reflect this fact.
```

The CHERIoT-Platform organisation is now a centralised landing pad for anything CHERIoT related.

The two repositories that have moved today are the ISA specification and the RTOS.
These are core parts of the platform.

> The [CHERIoT Sail repository](https://github.com/CHERIoT-Platform/cheriot-sail) contains the ISA specification, including both the formal model and the prose descriptions.

Reviewer: ```suggestion
The [CHERIoT Sail repository](https://github.com/CHERIoT-Platform/cheriot-sail) contains the ISA specification, including both the formal model and the [prose descriptions](https://cheriot-platform.github.io/cheriot-sail/cheriot-architecture.pdf).
```

Assuming that link works after the move.

This ecosystem is starting to grow.
We have merged improvements to the ISA specification and various parts of the software stack from Google.
SCI Semiconductor is working to ship commercial CHERIoT SoCs to customers next year.
lowRISC has built the [Sonata FPGA platform](https://www.sunburst-project.org) for prototyping CHERIoT devices.
Folks at [Configured Things](https://www.configuredthings.com) have written a [fantastic demo application](https://github.com/CHERIoT-Platform/cheriot-demos/tree/main/configuration_broker)

At this point, CHERIoT is no longer a research project, it is an open source foundation supported by multiple vendors and the repository moves reflect this fact.
The CHERIoT-Platform organisation is now a centralised landing pad for anything CHERIoT related.

The two repositories that have moved today are the ISA specification and the RTOS.
These are core parts of the platform.

The [CHERIoT Sail repository](https://github.com/CHERIoT-Platform/cheriot-sail) contains the ISA specification, including both the formal model and the prose descriptions.

> The formal model can be used to prove properties of the ISA, verify that implementations conform to the specification, verify properties of software running on CHERIoT cores, and also builds our golden model simulator.

Reviewer: ```suggestion
This executable formal model can be used to prove properties of the ISA, verify that implementations conform to the specification, verify properties of software running on CHERIoT cores, and also build our golden model simulator.
```

In addition to the Ibex, which aimed at production use, we also prototyped the CHERIoT ISA on the [CHERI Flute](https://github.com/CTSRD-CHERI/Flute) processor.
Ibex was optimised more for power and area, Flute more for performance.
Ibex was a 3-stage pipeline with a 33-bit memory bus (requiring two cycles to load a capability).
Flute was a 5-stage pipeline with a 64-bit memory bus (loading capabilities in a single cycle).

We did this to ensure that the CHERIoT ISA was not over-fitted for one particular microarchitecture.
We expect it to scale from (tiny!) things the size of Ibex up to quite large microcontrollers.

The Ibex, which vital to the ecosystem in demonstrating the viability of CHERIoT, was never intended to be the *only* CHERIoT core.
We welcome additional implementations (open and proprietary) and have no wish to bless any particular implementation and exclude others.

All of that said, the community engagement with CHERIoT Ibex has been amazing.
It has been the focus of several formal verification efforts and I am very happy to see contributors from across industry and academia working to improve it.

> We will continue to include a cycle-accurate simulator of the Ibex in the [CHERIoT Dev Container](https://github.com/orgs/CHERIoT-Platform/packages/container/package/devcontainer) and will add any other implementations that are available under license that permit redistribution.

Reviewer: ```suggestion
We will continue to include a cycle-accurate simulator of the Ibex in the [CHERIoT Dev Container](https://github.com/orgs/CHERIoT-Platform/packages/container/package/devcontainer) and will add any other implementations that are available under licenses that permit redistribution.
```
